### PR TITLE
*: split backup into BackupController and BackupManager.

### PR DESCRIPTION
**Pull request by fanminshi**
Previously, controlling logic and snapshot retrieving/storing logic are coupled together into a struct called backup.

This pr separates controlling logic and snapshot retrieving/storing logic out of backup into BackupController and BackupManager:

BackupController is responsible for controlling when to perform backup.
BackupManager is responsible for retrieving/storing backup.

### CODE REVIEW

**Comment by fanminshi** [comment]
cc/ @hongchaodeng @hasbro17

File: pkg/etcd-backup-operator/controller/sync.go [hunk 1]
```diff
@@ -90,3 +86,13 @@ func (b *Backup) handleErr(err error, key interface{}) {
 	// Report that, even after several retries, we could not successfully process this key
 	b.logger.Infof("Dropping Backup (%v) out of the queue: %v", key, err)
 }
+
+func (b *Backup) handleBackup(spec *api.EtcdBackupSpec) error {
```
Comment (hongchaodeng) [inline pkg/etcd-backup-operator/controller/sync.go]: Don't include this

File: pkg/cluster/backupstorage/s3.go [hunk 2]
```diff
@@ -37,7 +37,7 @@ func NewS3Storage(kubecli kubernetes.Interface, clusterName, ns string, p api.Ba
 		if err := os.MkdirAll(dir, 0700); err != nil {
 			return nil, err
 		}
-		options, err := setupAWSConfig(kubecli, ns, p.S3.AWSSecret, dir)
+		options, err := SetupAWSConfig(kubecli, ns, p.S3.AWSSecret, dir)
```
Comment (hongchaodeng) [inline pkg/cluster/backupstorage/s3.go]: Don't include change in this file in this PR.

File: pkg/backup/backup.go [hunk 3]
```diff
@@ -174,94 +176,168 @@ func (b *Backup) Run() {
 	}
 }
 
+// saveSnap saves the latest snapshot if the given current revision is greater lastSnapRev.
```
Comment (hongchaodeng) [inline pkg/backup/backup.go]: if latest revision is greater than the given lastSnapRev.

File: pkg/backup/backup.go [hunk 4]
```diff
@@ -174,94 +176,168 @@ func (b *Backup) Run() {
...
-		logrus.Warning(msg)
-		return lastSnapRev, fmt.Errorf(msg)
-	}
-	member, rev := getMemberWithMaxRev(pods, b.etcdTLSConfig, b.selfHosted)
-	if member == nil {
-		logrus.Warning("no reachable member")
-		return lastSnapRev, fmt.Errorf("no reachable member")
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultRequestTimeout)
+	snap, etcdVersion, rev, err := GetSnap(ctx, b.kclient, b.etcdTLSConfig, b.namespace, b.clusterName)
+	defer cancel()
```
Comment (hongchaodeng) [inline pkg/backup/backup.go]: Why defer? Just cancel here?

File: pkg/backup/backup.go [hunk 5]
```diff
@@ -174,94 +176,168 @@ func (b *Backup) Run() {
...
+	if err != nil {
+		return lastSnapRev, fmt.Errorf("failed to get snapshot: (%v)", err)
 	}
+	defer snap.Close()
```
Comment (hongchaodeng) [inline pkg/backup/backup.go]: Why defer? Why not just close after WriteSnap()?

File: pkg/backup/backup.go [hunk 6]
```diff
@@ -174,94 +176,168 @@ func (b *Backup) Run() {
...
+	defer rc.Close()
+	n, err := be.Save(version, rev, rc)
+	if err != nil {
+		return 0, err
+	}
+	return n, nil
+}
+
+// GetSnap gets the snapshot from the cluster member with the maximum revision.
+func GetSnap(ctx context.Context, kclient kubernetes.Interface, tls *tls.Config, namespace, clusterName string) (*Snap, string, int64, error) {
```
Comment (hongchaodeng) [inline pkg/backup/backup.go]: Why pass in a ctx here?

File: pkg/backup/backup.go [hunk 7]
```diff
@@ -174,94 +176,168 @@ func (b *Backup) Run() {
...
+		b.recentBackupsStatus = b.recentBackupsStatus[1:]
 	}
+
 	return rev, nil
 }
 
-func (b *Backup) writeSnap(m *etcdutil.Member, rev int64) error {
-	start := time.Now()
+// WriteSnap writes the snapshot to the backend.
+func WriteSnap(ctx context.Context, version string, rev int64, be backend.Backend, snap *Snap) (int64, error) {
```
Comment (hongchaodeng) [inline pkg/backup/backup.go]: Why pass in a ctx here?

File: pkg/backup/backup.go [hunk 8]
```diff
@@ -174,94 +176,168 @@ func (b *Backup) Run() {
...
-	ctx, cancel = context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
-	rc, err := etcdcli.Maintenance.Snapshot(ctx)
+func NewSnap(cfg clientv3.Config) *Snap {
+	return &Snap{cfg: cfg}
+}
+
+type Snap struct {
+	cfg clientv3.Config
+	// mu protects cli
+	mu  sync.Mutex
```
Comment (hongchaodeng) [inline pkg/backup/backup.go]: > // mu protects cli

Doesn't make sense. We don't need to protect cli in this code.

File: pkg/backup/backup.go [hunk 9]
```diff
@@ -174,94 +176,168 @@ func (b *Backup) Run() {
...
+	return NewSnap(cfg), resp.Version, rev, nil
+}
 
-	ctx, cancel = context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
-	rc, err := etcdcli.Maintenance.Snapshot(ctx)
+func NewSnap(cfg clientv3.Config) *Snap {
+	return &Snap{cfg: cfg}
+}
+
+type Snap struct {
```
Comment (hongchaodeng) [inline pkg/backup/backup.go]: This thing shouldn't call "Snap".
It should be SnapGetter or something.

Comment (hongchaodeng) [reply pkg/backup/backup.go]: If it's changed, same problem for GetSnap() and WriteSnap().

**Comment by fanminshi** [comment]
All fixed. PTAL cc/ @hongchaodeng

File: pkg/backup/backup.go [hunk 10]
```diff
@@ -174,94 +175,188 @@ func (b *Backup) Run() {
...
+
 	return rev, nil
 }
 
-func (b *Backup) writeSnap(m *etcdutil.Member, rev int64) error {
-	start := time.Now()
+// WriteSnap writes the snapshot to the backend.
+func WriteSnap(version string, rev int64, be backend.Backend, snap *Snap) (int64, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
+	defer cancel()
```
Comment (fanminshi) [inline pkg/backup/backup.go]: i used defer here so that `cancel` is called when both `rc, err := snap.Snapshot(ctx)` error out and `n, err := be.Save(version, rev, rc)` error out.

File: pkg/backup/backup.go [hunk 11]
```diff
@@ -174,94 +175,188 @@ func (b *Backup) Run() {
...
 func (b *Backup) saveSnap(lastSnapRev int64) (int64, error) {
-	podList, err := b.kclient.Core().Pods(b.namespace).List(k8sutil.ClusterListOpt(b.clusterName))
+	start := time.Now()
+
+	snap, rev, err := GetSnap(b.kclient, b.etcdTLSConfig, b.namespace, b.clusterName)
 	if err != nil {
-		return lastSnapRev, err
+		return lastSnapRev, fmt.Errorf("failed to get snapshot: (%v)", err)
 	}
+	defer snap.Close()
```
Comment (fanminshi) [inline pkg/backup/backup.go]: call `defer snap.Close()` here because I need to close snap when `if rev <= lastSnapRev {` is true, `etcdVersion, err := snap.Version()` errors out, and `n, err := WriteSnap(etcdVersion, rev, b.be, snap)` errors out.

File: pkg/backup/backup.go [hunk 12]
```diff
@@ -174,94 +175,186 @@ func (b *Backup) Run() {
...
 	if err != nil {
-		return fmt.Errorf("failed to create etcd client (%v)", err)
+		return nil, fmt.Errorf("failed to get member snapshot (%v)", err)
 	}
-	defer etcdcli.Close()
+	return rc, nil
+}
 
+// Version returns the etcd version of the etcd member.
+func (s *Snap) Version() (string, error) {
```
Comment (xiang90) [inline pkg/backup/backup.go]: this method does not make sense. why snapshot.version return member's version not snapshot's version? it seems too confusing.

File: pkg/backup/backup.go [hunk 13]
```diff
@@ -174,94 +175,186 @@ func (b *Backup) Run() {
...
+		return "", fmt.Errorf("failed to get member status (%v)", err)
 	}
+	return resp.Version, nil
+}
 
-	ctx, cancel = context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
-	rc, err := etcdcli.Maintenance.Snapshot(ctx)
-	if err != nil {
-		return fmt.Errorf("failed to receive snapshot (%v)", err)
+func (s *Snap) lazyClient() (*clientv3.Client, error) {
```
Comment (xiang90) [inline pkg/backup/backup.go]: do not expose implementation detail through method naming. 

also why client is a method on snap?

File: pkg/backup/backup.go [hunk 14]
```diff
@@ -174,94 +175,186 @@ func (b *Backup) Run() {
...
+		Endpoints:   []string{url},
 		DialTimeout: constants.DefaultDialTimeout,
-		TLS:         b.etcdTLSConfig,
+		TLS:         tls,
 	}
-	etcdcli, err := clientv3.New(cfg)
+	return &Snap{cfg: cfg}
+}
+
+type Snap struct {
```
Comment (xiang90) [inline pkg/backup/backup.go]: what does snap represent for? needs doc string

File: pkg/backup/backup.go [hunk 15]
```diff
@@ -174,94 +175,186 @@ func (b *Backup) Run() {
...
+
+	m, rev := getMemberWithMaxRev(pods, tls)
+	if m == nil {
+		return nil, 0, fmt.Errorf("no reachable member")
+	}
+
+	return NewSnap(m.ClientURL(), tls), rev, nil
+}
 
+func NewSnap(url string, tls *tls.Config) *Snap {
```
Comment (xiang90) [inline pkg/backup/backup.go]: doc string

File: pkg/backup/backup.go [hunk 16]
```diff
@@ -174,94 +175,186 @@ func (b *Backup) Run() {
...
+	n, err := be.Save(version, rev, rc)
+	rc.Close()
+	if err != nil {
+		return 0, err
+	}
+	return n, nil
+}
+
+// GetSnap gets the snapshot from the cluster member with the maximum revision.
+func GetSnap(kclient kubernetes.Interface, tls *tls.Config, namespace, clusterName string) (*Snap, int64, error) {
```
Comment (xiang90) [inline pkg/backup/backup.go]: does this func really return a snapshot?

File: pkg/backup/backup.go [hunk 17]
```diff
@@ -174,94 +175,186 @@ func (b *Backup) Run() {
 	}
 }
 
+// saveSnap saves the latest snapshot if latest revision is greater than the given lastSnapRev.
 func (b *Backup) saveSnap(lastSnapRev int64) (int64, error) {
```
Comment (xiang90) [inline pkg/backup/backup.go]: probably should rename Backup to BackupManager to be more descriptive during this refactor.

File: pkg/backup/backup.go (hunk 12 above)
Comment (fanminshi) [reply pkg/backup/backup.go]: Maybe change method name to `EtcdVersion()` or remove`Version()` in side snap struct?

Comment (xiang90) [reply pkg/backup/backup.go]: probably https://github.com/coreos/etcd-operator/pull/1489#discussion_r143321802 needs to be answered first.

File: pkg/backup/backup.go (hunk 13 above)
Comment (fanminshi) [reply pkg/backup/backup.go]: `lazyClient()` is use to create client lazily and is used in both `Version()` and `Snapshot()`. So I refactor the common client initialization code as `lazyClient()`.

File: pkg/backup/backup.go (hunk 16 above)
Comment (fanminshi) [reply pkg/backup/backup.go]: not anymore. will change the doc string

File: pkg/backup/backup.go (hunk 13 above)
Comment (xiang90) [reply pkg/backup/backup.go]: i do not think the snap struct makes a lot of sense to me at the first place.

things are getting too confusing here.

first backup is the controller that manages backups. to make it less confusing, the actual backup is called snapshot.

now we are introducing two similar things again: snap and snapshot.

a backup can get a snap to get snapshots... it is just getting super confusing.

File: pkg/backup/backup.go (hunk 17 above)
Comment (fanminshi) [reply pkg/backup/backup.go]: kk.

**Comment by xiang90** [comment]
@fanminshi 

Can you write down the motivation of the refactor? What do we want to achieve? Why do we want to achieve? It helps me to understand the direction of this PR. thanks.

**Comment by fanminshi** [comment]
@xiang90 kk, I'll write the motivation in the commit msg.

File: pkg/backup/backup.go (hunk 14 above)
Comment (fanminshi) [reply pkg/backup/backup.go]: Snap provides an abstraction of getting snapshot from a member. 
 so calling snap.snapshot() returns the snapshot. calling snap.close() releases resources which is the etcd client.

Comment (xiang90) [reply pkg/backup/backup.go]: Then it should be named as SnapshotGetter. I am not sure if this abstraction is really necessary though.

**Comment by fanminshi** [comment]
@xiang90 I added motivation in the commit msg section. let me know if you want me to be more detailed.

File: pkg/backup/backup.go (hunk 14 above)
Comment (fanminshi) [reply pkg/backup/backup.go]: let me rethink about this.

**Comment by xiang90** [comment]
> They needs two functionalities:
> get snapshot from etcd cluster
> save snapshot to the backend.

Basically, previously we couple the controlling logic and snapshot retrieving/storing logic together into a struct called backup.

We should do here are:

1. rename backup to backupController to make it clear that it is responsible for the controlling logic (a run func to drive the for loop)

2. move everything else to backupManager struct.

it should look like

```go

func (bc *BackupController) Run() {
	go bc.startHTTP()

	lastSnapRev := bc.backupManager.getLatestBackupRev()
	interval := constants.DefaultSnapshotInterval
	if bc.policy.BackupIntervalInSecond != 0 {
		interval = time.Duration(b.policy.BackupIntervalInSecond) * time.Second
	}

	go func() {
		if bc.policy.MaxBackups == 0 {
			return
		}
		for {
			<-time.After(10 * time.Second)
			err := bc.backupManager.purge(b.policy.MaxBackups)
			if err != nil {
				logrus.Errorf("fail to purge backups: %v", err)
			}
		}
	}()

	for {
		var ackchan chan backupNowAck
		select {
		case <-time.After(interval):
		case ackchan = <-b.backupNow:
			logrus.Info("received a backup request")
		}

		rev, err := bc.backupManager.saveSnap(lastSnapRev)
		if err != nil {
			logrus.Errorf("failed to save snapshot: %v", err)
		}
		lastSnapRev = rev

		if ackchan != nil {
			ack := backupNowAck{err: err}
			if err == nil {
				ack.status = b.backupManager.getLatestBackupStatus()
			}
			ackchan <- ack
		}
	}
}
```

**Comment by fanminshi** [comment]
@xiang90 I get rid of GetSnap() and WriteSnap() and collapse both into SaveSnap() function where  SaveSnap() find snapshot and save it to the backend.

I also refactor the  backup -> backup controller and backup manager where backup controller when to backup and  backup manager  manages how to backup.

File: pkg/backup/backup.go [hunk 18]
```diff
@@ -167,101 +174,128 @@ func (b *Backup) Run() {
...
 	}
+	return rev
+}
+
+func (b *BackupManager) getLatestBackupStatus() backupapi.BackupStatus {
+	return b.recentBackupsStatus[len(b.recentBackupsStatus)-1]
+}
 
-	return nil
+func getRunningPods(kclient kubernetes.Interface, namespace, clusterName string) ([]*v1.Pod, error) {
```
Comment (xiang90) [inline pkg/backup/backup.go]: revert this change? is this relevant to what we want to achieve?

File: pkg/backup/backup.go [hunk 19]
```diff
@@ -298,23 +332,3 @@ func getMemberWithMaxRev(pods []*v1.Pod, tc *tls.Config, selfHosted bool) (*etcd
 	}
 	return member, maxRev
 }
-
-func (b *Backup) getLatestBackupRev() int64 {
```
Comment (xiang90) [inline pkg/backup/backup.go]: do not move this around. just change the receiver to backupmanager to ensure minimal change.

File: pkg/backup/backup.go [hunk 20]
```diff
@@ -298,23 +332,3 @@ func getMemberWithMaxRev(pods []*v1.Pod, tc *tls.Config, selfHosted bool) (*etcd
...
-		return 0
-	}
-	rev, err := util.GetRev(name)
-	if err != nil {
-		logrus.Fatal(err)
-	}
-	return rev
-}
-
-func (b *Backup) getLatestBackupStatus() backupapi.BackupStatus {
```
Comment (xiang90) [inline pkg/backup/backup.go]: same here.

**Comment by xiang90** [comment]
we almost get there. thanks.

File: pkg/backup/backup.go (hunk 18 above)
Comment (fanminshi) [reply pkg/backup/backup.go]: not anymore. will revert.

File: pkg/backup/backup.go [hunk 21]
```diff
@@ -46,25 +45,30 @@ const (
 	maxRecentBackupStatusCount = 10
 )
 
-type Backup struct {
+// BackupManager manages how to backup.
```
Comment (xiang90) [inline pkg/backup/backup.go]: what does how mean? need to be more descriptive.

File: pkg/backup/backup.go [hunk 22]
```diff
@@ -114,36 +118,39 @@ func New(kclient kubernetes.Interface, clusterName, ns string, sp api.ClusterSpe
...
+		listenAddr:    listenAddr,
+		backupNow:     make(chan chan backupNowAck),
+		policy:        *sp.Backup,
+		backupManager: bm,
 	}, nil
 }
 
-func (b *Backup) Run() {
-	go b.startHTTP()
+func (bc *BackupController) Run() {
```
Comment (xiang90) [inline pkg/backup/backup.go]: doc string on run

File: pkg/backup/backup.go [hunk 23]
```diff
@@ -167,101 +174,128 @@ func (b *Backup) Run() {
...
-	log.Printf("saving backup for cluster (%s)", b.clusterName)
-	if err := b.writeSnap(member, rev); err != nil {
-		err = fmt.Errorf("write snapshot failed: %v", err)
-		return lastSnapRev, err
+		return 0, 0, "", fmt.Errorf("skipped creating new backup: no change since last time")
 	}
-	return rev, nil
-}
-
-func (b *Backup) writeSnap(m *etcdutil.Member, rev int64) error {
```
Comment (xiang90) [inline pkg/backup/backup.go]: why this func is removed? only the receiver needs to be changed to backupManager to achieve our goal, no?

File: pkg/backup/backup.go [hunk 24]
```diff
@@ -46,25 +45,30 @@ const (
...
-func New(kclient kubernetes.Interface, clusterName, ns string, sp api.ClusterSpec, listenAddr string) (*Backup, error) {
+// BackupController controls when to backup.
+type BackupController struct {
+	listenAddr    string
+	backupNow     chan chan backupNowAck
+	policy        api.BackupPolicy
+	backupManager *BackupManager
+}
+
+func New(kclient kubernetes.Interface, clusterName, ns string, sp api.ClusterSpec, listenAddr string) (*BackupController, error) {
```
Comment (xiang90) [inline pkg/backup/backup.go]: NewBackupController

File: pkg/backup/backup.go [hunk 25]
```diff
@@ -167,101 +174,128 @@ func (b *Backup) Run() {
...
+	b.recentBackupsStatus = append(b.recentBackupsStatus, bs)
+	if len(b.recentBackupsStatus) > maxRecentBackupStatusCount {
+		b.recentBackupsStatus = b.recentBackupsStatus[1:]
+	}
+
+	return rev, nil
+}
 
+// SaveSnap saves snapshot to backend from the etcd member whose rev is greater than lastSnapRev.
+func SaveSnap(kclient kubernetes.Interface, tls *tls.Config, lastSnapRev int64, be backend.Backend, namespace, clusterName string) (int64, int64, string, error) {
```
Comment (xiang90) [inline pkg/backup/backup.go]: why do we need to create this func there? we should expose backupManager and its methods.

File: pkg/backup/backup.go (hunk 21 above)
Comment (fanminshi) [reply pkg/backup/backup.go]: `BackupManager manages where to get etcd backups from and where save them`?

File: pkg/backup/backup.go (hunk 23 above)
Comment (xiang90) [reply pkg/backup/backup.go]: /cc @fanminshi  this needs to be resolved

File: pkg/backup/backup.go (hunk 25 above)
Comment (xiang90) [reply pkg/backup/backup.go]: /cc @fanminshi this needs to be resolved.

Comment (fanminshi) [reply pkg/backup/backup.go]: kk, let me see what I can do.

File: pkg/backup/backup.go (hunk 23 above)
Comment (fanminshi) [reply pkg/backup/backup.go]: `only the receiver needs to be changed to backupManager to achieve our goal, no?`
possibly. let me think about how to use `backupManager` in the operator controller code.

File: pkg/backup/backup.go [hunk 26]
```diff
@@ -154,30 +162,38 @@ func (b *Backup) Run() {
...
+				ack.status = bc.recentBackupsStatus[len(bc.recentBackupsStatus)-1]
 			}
 			ackchan <- ack
 		}
 	}
 }
 
-func (b *Backup) saveSnap(lastSnapRev int64) (int64, error) {
-	podList, err := b.kclient.Core().Pods(b.namespace).List(k8sutil.ClusterListOpt(b.clusterName))
+// SaveSnap saves the snapshot if snapshot's revision is greater than the given lastSnapRev.
```
Comment (xiang90) [inline pkg/backup/backup.go]: explain the return value.

Comment (xiang90) [reply pkg/backup/backup.go]: also does backupstate contain rev already?

**Comment by xiang90** [comment]
lgtm after fixing the two minor issues.

File: pkg/backup/backup.go (hunk 26 above)
Comment (fanminshi) [reply pkg/backup/backup.go]: will do what you suggested.

File: pkg/backup/backup.go [hunk 27]
```diff
@@ -154,30 +162,39 @@ func (b *Backup) Run() {
...
 
-		rev, err := b.saveSnap(lastSnapRev)
+		bs, err := bc.backupManager.SaveSnap(lastSnapRev)
 		if err != nil {
 			logrus.Errorf("failed to save snapshot: %v", err)
 		}
-		lastSnapRev = rev
+
+		if bs != nil {
+			lastSnapRev = bs.Revision
```
Comment (fanminshi) [inline pkg/backup/backup.go]: @xiang90 I changed `SaveSnap` to return `BackupStatus` and `error`.
And only updates lastSnapRev if `BackupStatus` is not nil.

File: pkg/backup/backup.go [hunk 28]
```diff
@@ -191,77 +208,74 @@ func (b *Backup) saveSnap(lastSnapRev int64) (int64, error) {
...
 	if member == nil {
 		logrus.Warning("no reachable member")
-		return lastSnapRev, fmt.Errorf("no reachable member")
+		return nil, fmt.Errorf("no reachable member")
 	}
 
 	if rev <= lastSnapRev {
 		logrus.Info("skipped creating new backup: no change since last time")
-		return lastSnapRev, nil
+		return nil, nil
```
Comment (fanminshi) [inline pkg/backup/backup.go]: @xiang90 I am unsure what to return here. I return `nil,nil` because no backup is performed and skipping is not considered an error by judgement of previous code.

**Comment by xiang90** [comment]
lgtm

**Comment by fanminshi** [comment]
merge when green.
